### PR TITLE
genesis: configurable blob size

### PR DESCRIPTION
**Pull request by ibhatt-jumptrading**

### CODE REVIEW

File: src/discof/genesis/fd_genesi_tile.c [hunk 1]
```diff
@@ -251,14 +254,19 @@ after_credit( fd_genesi_tile_t *  ctx,
     int bzerr = BZ2_bzDecompressInit( &bzstrm, 0, 0 );
     if( FD_UNLIKELY( BZ_OK!=bzerr ) ) FD_LOG_ERR(( "BZ2_bzDecompressInit() failed (%d)", bzerr ));
 
-    ulong decompressed_sz = FD_GENESIS_MAX_MESSAGE_SIZE;
+    ulong decompressed_sz = ctx->max_message_size + 4UL*512UL;
```
Comment (ripatel-fd) [inline src/discof/genesis/fd_genesi_tile.c]: why 4UL*512UL? is this for the tar headers?

File: src/flamenco/genesis/fd_genesis_parse.h [hunk 2]
```diff
@@ -3,7 +3,8 @@
 
 #include "../fd_flamenco_base.h"
 
-/* Hardcoded max serialized genesis blob size */
+/* Hard safety ceiling for the serialized genesis blob size.  The
+   runtime-configured limit must not exceed this value. */
 #define FD_GENESIS_MAX_MESSAGE_SIZE (1UL<<28) /* 256 MiB */
```
Comment (ripatel-fd) [inline src/flamenco/genesis/fd_genesis_parse.h]: can we not just make this 2GB or something, so we never run into the issue again?

**Review by ripatel-fd — DISMISSED** [review]

File: src/discof/genesis/fd_genesi_tile.c [hunk 3]
```diff
@@ -356,12 +364,17 @@ process_local_genesis( fd_genesi_tile_t * ctx,
...
   ctx->genesis_blob_sz = 0UL;
   for(;;) {
-    if( FD_UNLIKELY( ctx->genesis_blob_sz>=FD_GENESIS_MAX_MESSAGE_SIZE ) ) {
-      FD_LOG_ERR(( "The genesis file at `%s` is too large for this Firedancer build.\n"
-                   "Cannot start Firedancer. Please use a different genesis config or increase FD_GENESIS_MAX_MESSAGE_SIZE.",
-                   genesis_path ));
+    if( FD_UNLIKELY( ctx->genesis_blob_sz==ctx->max_message_size ) ) {
+      uchar extra;
+      long result = read( ctx->in_fd, &extra, 1UL );
+      if( FD_UNLIKELY( -1==result ) ) FD_LOG_ERR(( "read() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/discof/genesis/fd_genesi_tile.c:370]: You can ignore this this is nonsense

File: src/app/firedancer/config/default.toml [hunk 4]
```diff
@@ -725,6 +725,14 @@ telemetry = true
     # or test environments to decrease memory usage or test edge cases.
     max_fork_width = 32
 
+    # Limits applied while loading the serialized genesis configuration.
+    # Larger values increase the memory reserved by the genesis and RPC
+    # tiles, as well as the genesi_out link.
+    [runtime.genesis]
+        # Maximum size in MiB of the uncompressed genesis.bin file, for
+        # both local files and downloaded archives.  Must be in [1, 256].
+        max_message_size_mib = 10
```
[the remark this answers is not in the document]
Comment (ripatel-fd) [reply src/app/firedancer/config/default.toml]: seems valid

File: src/app/firedancer/config/default.toml [hunk 5]
```diff
@@ -725,6 +725,14 @@ telemetry = true
     # or test environments to decrease memory usage or test edge cases.
     max_fork_width = 32
 
+    # Limits applied while loading the serialized genesis configuration.
+    # Larger values increase the memory reserved by the genesis and RPC
+    # tiles, as well as the genesi_out link.
+    [runtime.genesis]
```
Comment (mmcgee-jump) [inline src/app/firedancer/config/default.toml]: I don't really think this should be configurable? This is not a useful knob for an operator. If you really need to, put it under development section.

**Review by mmcgee-jump — CHANGES_REQUESTED** [review]
.

File: src/app/firedancer/config/default.toml (hunk 5 above)
Comment (ibhatt-jumptrading) [reply src/app/firedancer/config/default.toml]: agave exposes this as a CLI flag so I wanted to match that behavior. moving to development seems reasonable

File: src/discof/rpc/fd_rpc_tile.c [hunk 6]
```diff
@@ -2399,7 +2411,7 @@ privileged_init( fd_topo_t const *      topo,
   fd_rpc_tile_t * ctx      = FD_SCRATCH_ALLOC_APPEND( l, alignof( fd_rpc_tile_t ), sizeof( fd_rpc_tile_t ) );
   fd_http_server_t * _http = FD_SCRATCH_ALLOC_APPEND( l, fd_http_server_align(),   fd_http_server_footprint( http_params ) );
 
-  fd_memset( ctx, 0, offsetof(fd_rpc_tile_t, genesis_tar) ); /* trailing large buffers are write-before-read scratch */
+  fd_memset( ctx, 0, sizeof(fd_rpc_tile_t) );
```
Comment (mmcgee-jump) [inline src/discof/rpc/fd_rpc_tile.c:2414]: Why?

File: src/app/firedancer/config/default.toml [hunk 7]
```diff
@@ -1927,6 +1927,11 @@ telemetry = true
         # genesis hash.
         validate_genesis_hash = true
 
+        # The maximum size of the genesis file that is either loaded
+        # from the validator's host or downloaded from the network.
+        # Must be in [1, 4055].
+        max_message_size_mib = 10
```
Comment (mmcgee-jump) [inline src/app/firedancer/config/default.toml]: This is a pretty weird name, what is a message here? Probably `max_file_size_mib` is better?

File: src/discof/rpc/fd_rpc_tile.c (hunk 6 above)
Comment (ibhatt-jumptrading) [reply src/discof/rpc/fd_rpc_tile.c:2414]: size of struct is small now, can safely just zero out instead of the partial zeroing

File: src/app/firedancer/config/default.toml [hunk 8]
```diff
@@ -1927,6 +1927,11 @@ telemetry = true
         # genesis hash.
         validate_genesis_hash = true
 
+        # The maximum size of the genesis file that is either loaded
+        # from the validator's host or downloaded from the network.
```
Comment (mmcgee-jump) [inline src/app/firedancer/config/default.toml]:
```suggestion
        # from disk or downloaded from the network.
```

**Review by mmcgee-jump — DISMISSED** [review]

**Review by mmcgee-jump — APPROVED** [review]

**Review by yufeng-jump — APPROVED** [review]
